File: src/main/java/alfheim/common/spell/darkness/SpellPoisonRoots.java
package alfheim.common.spell.darkness;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.math.Vector3;
import alfheim.api.entity.EnumRace;
import alfheim.api.spell.SpellBase;
import alfheim.api.spell.SpellBase.SpellCastResult;
import alfheim.common.core.handler.CardinalSystem.PartySystem;
import alfheim.common.core.handler.CardinalSystem.PartySystem.Party;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;

public class SpellPoisonRoots extends SpellBase {

	@Override
	public SpellCastResult performCast(EntityLivingBase caster) {
		Party pt = caster instanceof EntityPlayer ? PartySystem.getParty((EntityPlayer) caster) : PartySystem.getMobParty(caster);
		if (pt == null) return SpellCastResult.NOTARGET;
		boolean flagBadEffs = false, flagNotParty = false;
		EntityLivingBase member;	
		
		scanpt: for (int i = 0; i < pt.count; i++) {
			member = pt.get(i);
			if (member == null || Vector3.entityDistance(caster, member) > 32) continue;
			for (Object o : member.getActivePotionEffects()) {
				if (Potion.potionTypes[((PotionEffect) o).getPotionID()].isBadEffect) {
					flagBadEffs = true;
					break scanpt;
				}
			}
		}
		
		if (!flagBadEffs) return SpellCastResult.WRONGTGT;
		
		List<EntityLivingBase> l = caster.worldObj.getEntitiesWithinAABB(EntityLivingBase.class, caster.boundingBox.expand(16, 16, 16));
		for (EntityLivingBase e : l) {
			if (pt.isMember(e)) {
				flagNotParty = true;
				break;
			}
		}
		
		if (!flagNotParty) return SpellCastResult.NOTARGET;
		
		SpellCastResult result = checkCast(caster);
		if (result != SpellCastResult.OK) return result;
		
		ArrayList<PotionEffect> remove = new ArrayList<PotionEffect>();
		Iterator<EntityLivingBase> mobs = l.iterator();
		EntityLivingBase target = mobs.next();
		PotionEffect pe;
		
		for (int i = 0; i < pt.count; i++) {
			member = pt.get(i);
			for (Object o : member.getActivePotionEffects()) {
				pe = (PotionEffect) o;
				
				while (pt.isMember(target) && mobs.hasNext()) target = mobs.next();
				if (pt.isMember(target)) return SpellCastResult.NOTARGET;			// Some desync, sorry for your mana :(
				
				if (Potion.potionTypes[pe.getPotionID()].isBadEffect) {
					target.addPotionEffect(new PotionEffect(pe.potionID, pe.duration, pe.amplifier, pe.isAmbient));
					remove.add(pe);
					if (mobs.hasNext()) target = mobs.next();
					else mobs = l.iterator();
				}
			}
			
			for (PotionEffect r : remove) member.removePotionEffect(r.potionID);
			remove.clear();
		}
		
		for (EntityLivingBase e : l) if (!pt.isMember(e)) e.addPotionEffect(new PotionEffect(Potion.moveSlowdown.id, 300, 4, true));
		
		return result;
	}

	@Override
	public EnumRace getRace() {
		return EnumRace.IMP;
	}

	@Override
	public String getName() {
		return "poisonroots";
	}

	@Override
	public int getManaCost() {
		return 60000;
	}

	@Override
	public int getCooldown() {
		return 6000;
	}

	@Override
	public int castTime() {
		return 30;
	}

	@Override
	public boolean isHard() {
		return false;
	}

	@Override
	public void render(EntityLivingBase caster) {}
}

File: src/main/java/alfheim/common/spell/darkness/SpellPoisonRoots.kt
package alfheim.common.spell.darkness

import alexsocol.asjlib.math.Vector3
import alfheim.api.entity.EnumRace
import alfheim.api.spell.SpellBase
import alfheim.common.core.handler.CardinalSystem.PartySystem
import alfheim.common.core.util.*
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.potion.*
import java.util.*

object SpellPoisonRoots: SpellBase("poisonroots", EnumRace.IMP, 60000, 6000, 30) {
	
	override var duration = 300
	override var efficiency = 4.0
	
	override val usableParams
		get() = arrayOf(duration, efficiency, radius)
	
	override fun performCast(caster: EntityLivingBase): SpellCastResult {
		val pt = (if (caster is EntityPlayer) PartySystem.getParty(caster) else PartySystem.getMobParty(caster)) ?: return SpellCastResult.NOTARGET
		var flagBadEffs = false
		var member: EntityLivingBase?
		
		scanpt@ for (i in 0 until pt.count) {
			member = pt[i]
			if (member == null || Vector3.entityDistance(caster, member) > 32) continue
			for (o in member.activePotionEffects) {
				if (Potion.potionTypes[(o as PotionEffect).getPotionID()].isBadEffect) {
					flagBadEffs = true
					break@scanpt
				}
			}
		}
		
		if (!flagBadEffs) return SpellCastResult.WRONGTGT
		
		val l = caster.worldObj.getEntitiesWithinAABB(EntityLivingBase::class.java, caster.boundingBox.expand(radius)) as List<EntityLivingBase>
		val flagNotParty = l.any { pt.isMember(it) }
		
		if (!flagNotParty) return SpellCastResult.NOTARGET
		
		val result = checkCast(caster)
		if (result != SpellCastResult.OK) return result
		
		val remove = ArrayList<PotionEffect>()
		var mobs = l.iterator()
		var target = mobs.next()
		var pe: PotionEffect
		
		for (i in 0 until pt.count) {
			member = pt[i] ?: continue
			for (o in member.activePotionEffects) {
				pe = o as PotionEffect
				
				while (pt.isMember(target) && mobs.hasNext()) target = mobs.next()
				if (pt.isMember(target)) return SpellCastResult.NOTARGET            // Some desync, sorry for your mana :(
				
				if (Potion.potionTypes[pe.getPotionID()].isBadEffect) {
					target.addPotionEffect(PotionEffect(pe.potionID, pe.duration, pe.amplifier, pe.isAmbient))
					remove.add(pe)
					if (mobs.hasNext())
						target = mobs.next()
					else
						mobs = l.iterator()
				}
			}
			
			for (r in remove) member.removePotionEffect(r.potionID)
			remove.clear()
		}
		
		for (e in l) if (!pt.isMember(e)) e.addPotionEffect(PotionEffect(Potion.moveSlowdown.id, duration, efficiency.I, true))
		
		return result
	}
}